server/schedulers/utils.go

```diff
@@ -58,18 +58,31 @@ func shouldBalance(cluster schedule.Cluster, source, target *core.StoreInfo, reg
 	// The reason we use max(regionSize, averageRegionSize) to check is:
 	// 1. prevent moving small regions between stores with close scores, leading to unnecessary balance.
 	// 2. prevent moving huge regions, leading to over balance.
-	regionSize := region.GetApproximateSize()
-	if regionSize < cluster.GetAverageRegionSize() {
-		regionSize = cluster.GetAverageRegionSize()
-	}
-
-	regionSize = int64(float64(regionSize) * adjustTolerantRatio(cluster))
-	sourceDelta := opInfluence.GetStoreInfluence(source.GetID()).ResourceSize(kind) - regionSize
-	targetDelta := opInfluence.GetStoreInfluence(target.GetID()).ResourceSize(kind) + regionSize
+	leaderScheduleKind := cluster.GetLeaderScheduleKind()
+	tolerantResource := getTolerantResource(cluster, region, leaderScheduleKind)
+	sourceDelta := opInfluence.GetStoreInfluence(source.GetID()).ResourceSize(kind, leaderScheduleKind) - tolerantResource
+	targetDelta := opInfluence.GetStoreInfluence(target.GetID()).ResourceSize(kind, leaderScheduleKind) + tolerantResource
 
 	// Make sure after move, source score is still greater than target score.
-	return source.ResourceScore(kind, cluster.GetHighSpaceRatio(), cluster.GetLowSpaceRatio(), sourceDelta) >
-		target.ResourceScore(kind, cluster.GetHighSpaceRatio(), cluster.GetLowSpaceRatio(), targetDelta)
+	return source.ResourceScore(kind, cluster.GetHighSpaceRatio(), cluster.GetLowSpaceRatio(), sourceDelta, leaderScheduleKind) >
+		target.ResourceScore(kind, cluster.GetHighSpaceRatio(), cluster.GetLowSpaceRatio(), targetDelta, leaderScheduleKind)
+}
+
+func getTolerantResource(cluster schedule.Cluster, region *core.RegionInfo, leaderScheduleKind core.LeaderScheduleKind) int64 {
+	switch leaderScheduleKind {
+	case core.ScheduleLeaderByCount:
+		leaderCount := int64(float64(1) * adjustTolerantRatio(cluster))
```

Comment:
If we schedule leader by count, do we still need the tolerant mechanism?

Reply:
yes, we do.